init commit

derren-korean opened the pull request:
풀리퀘스트 테스트.

derren-korean committed:
init commit

derren-korean committed:
step1 commit

javajigi commented on src/main/java/racing/Car.java:
```diff
@@ -0,0 +1,29 @@
+package racing;
+
+public class Car {
+    private int number;
```
number 인스턴스는 굳이 필요 없을 것 같음.

javajigi commented on src/main/java/racing/Car.java:
```diff
@@ -0,0 +1,29 @@
+        return number;
+    }
+
+    public void setNumber(int number) {
```
setter 메소드를 사용하지 않는다.

javajigi commented on src/main/java/racing/Race.java:
```diff
@@ -0,0 +1,139 @@
+    private final int MIN_COUNT = 1;
+    private final int MAX_COUNT = 100;
+    private final int MIN_FORWARD = 4;
+    private final int RANDOM_BOUND = 10;
```
상수 값은 'static final'을 쓰는 것이 맞음.

javajigi commented on src/main/java/racing/Race.java:
```diff
@@ -0,0 +1,139 @@
+    public Race() {
+    }
+
+    public int getCarCount() {
```
인스턴스 변수가 있으면 습관적으로 setter/getter 만드는 것으로 보여짐.
정말 필요한 경우가 아니면 setter/getter를 만들지 않는다.

javajigi commented on src/main/java/racing/Race.java:
```diff
@@ -0,0 +1,139 @@
+        return Integer.parseInt(temp);
+    }
+
+    private boolean isOutOfNumber(String temp) {
```
에러 체크 메소드들 정말 잘했네요. 💯 
그런데 이 코드에 대한 단위 테스트 코드까지 있으면 더 좋겠음.

javajigi commented on src/main/java/racing/Race.java:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * UI 전용.
+     */
+    @Deprecated
```
@Deprecated을 추가한 이유는?

javajigi commented on src/main/java/racing/Race.java:
```diff
@@ -0,0 +1,139 @@
+     * UI 전용.
+     */
+    @Deprecated
+    private void setCountByUser() {
```
UI 구현부와 로직을 구현하는 부분은 같은 객체에 존재하고 있음.
두 부분을 서로 다른 클래스로 분리할 것을 추천.

javajigi commented on src/main/java/racing/Race.java:
```diff
@@ -0,0 +1,139 @@
+
+    void enrollRacingCars() {
+        if (this.carCount == 0) throw new RuntimeException("carCount가 0 입니다. initRace 메소드를 실행 후 진행하세요.");
+        List<Car> carList = new ArrayList<>();
```
list와 같은 자료 구조를 변수 이름을 사용하지 않을 것을 추천.
이 경우 cars와 같은 복수형을 사용

javajigi reviewed (changes requested):
전체적으로 예외처리까지 고려해서 구현 잘했어요. 그런데 setter/getter를 즐겨 사용하네요. 앞으로 가능하면 이 둘을 멀리해 보면 좋을 것 같아요.

derren-korean committed:
refactor : Car.java number제거,
상수값 static으로 선언,
UI 객체 RaceVO로 수정,
단위 테스트 추가(UI입력 문자 판단 로직, 예외 처리 테스트),
deprecated 애노테이션 제거(step1에서는 UI로직 필요 없을 것 같아 애노테이션 처리 했었습니다.),
같은 자료 구조를 변수이름으로 사용한것 변경.(carList -> cars)

javajigi commented on src/main/java/racing/Car.java:
```diff
@@ -0,0 +1,16 @@
+        return distance;
+    }
+
+    public void move() {
```
random 값에 따라 이동 여부를 결정하는 로직을 이 메소드에서 구현하면 좋겠네요.

javajigi commented on src/main/java/racing/RaceVO.java:
```diff
@@ -0,0 +1,42 @@
+    public void enrollCars() {
+        if (getCarCount() == 0) throw new RuntimeException("carCount가 0 입니다. initRace 메소드를 실행 후 진행하세요.");
+        List<Car> carList = new ArrayList<>();
+        for (int i = 0; i < getCarCount(); i++) {
```
인스턴스 변수로 있다면 굳이 getCarCount()가 아니라 carCount로 접근하면 좋겠어요.

javajigi commented on src/main/java/racing/RaceVO.java:
```diff
@@ -0,0 +1,42 @@
+    private int maxMoveCount;
+    private List<Car> cars;
+
+    public RaceVO() {
```
carCount, maxMoveCount에 대해 setter 메소드를 사용하지 말고 이 생성자의 인자로 전달해 보세요.

javajigi commented on src/main/java/racing/RaceVO.java:
```diff
@@ -0,0 +1,42 @@
+    private List<Car> cars;
+
+    public RaceVO() {
+    }
```
생성자에서 carCount를 받은 후 List<Car>를 바로 생성해 보세요.

javajigi commented on src/main/java/racing/Race.java:
```diff
@@ -0,0 +1,108 @@
+        initRace(0,0);
+    }
+
+    public void initRace(int carCount, int moveCount) {
```
RaceVO를 다음과 같이 구현해 보세요.

```
int carCount = InputView.getCarCount();
int maxMoveCount = InputView.getMaxMoveCount();

vo = new RaceVO(carCount, maxMoveCount);
```

javajigi reviewed (changes requested):
클래스의 역할을 좀 더 분리하고 setter/getter 메소드 사용을 좀 자제한다면 코드가 훨씬 더 깔금하고 안전한 코드가 될 것 같아요.

derren-korean committed:
refactor : Car.java 레이싱카 이외의 다른 차는 동작이 다를 수 있기 때문에, 확장성을 위해 interface로 구현, InputView.java final과 static의 용도로 공부하면서... 조잡하지만 시도해보았습니다. Race.java->RaceGame.java 명칭 변경 및 기능 제거, RaceVO.java setter/getter를 제거한 형태로 생성자의 인자로 코드 간략화 RacingCar.java move기능 강화

javajigi commented on src/main/java/racing/RacingCar.java:
```diff
@@ -0,0 +1,30 @@
+        for (int i = 0; i < tryCount; i++) {
+            if (getRandom() < MIN_FORWARD_NUMBER) continue;
+            this.distance++;
+        }
```
이 메소드는 random 값 때문에 여전히 test하기 힘드네요.
다음과 같은 구조로 변경해 구현하고 테스트 해보세요.

```
public void move(final int randomValue) {
   // 구현
}
```

javajigi reviewed (changes requested):
구현 잘 하셨네요. 핵심 로직에 대한 테스트를 추가했으면 하는 바람으로 피드백 하나 남겨요.

derren-korean committed:
refactor : RacingCar객체의 move메소드 변경, test가 가능한 구조로 구현 및 테스트
